modules/wallet/scan.go

```diff
@@ -0,0 +1,198 @@
+	for i, k := range generateKeys(s.seed, initialProgress, n) {
+		s.keys[k.UnlockConditions.UnlockHash()] = initialProgress + uint64(i)
+	}
+}
+
+// ProcessConsensusChange scans the blockchain for information relevant to the
+// seedScanner.
+func (s *seedScanner) ProcessConsensusChange(cc modules.ConsensusChange) {
+	s.blockheight -= types.BlockHeight(len(cc.RevertedBlocks))
+	// update outputs
+	for _, block := range cc.AppliedBlocks {
+		s.blockheight++
+		// when a miner payout matures, move it from minerOutputs to
+		// siacoinOutputs
+		if matured, exists := s.minerOutputs[s.blockheight]; exists {
+			for _, maturedOutput := range matured {
+				s.siacoinOutputs[maturedOutput.id] = maturedOutput
+			}
+			delete(s.minerOutputs, s.blockheight)
+		}
+		for i, mp := range block.MinerPayouts {
+			// if a seed miner output is found, add it to the delayed output
+			// set
+			if index, exists := s.keys[mp.UnlockHash]; exists && mp.Value.Cmp(s.dustThreshold) > 0 {
+				maturityHeight := s.blockheight + types.MaturityDelay
+				s.minerOutputs[maturityHeight] = append(s.minerOutputs[maturityHeight], scannedOutput{
+					id:        types.OutputID(block.MinerPayoutID(uint64(i))),
+					value:     mp.Value,
+					seedIndex: index,
+				})
+			}
+		}
+		for _, txn := range block.Transactions {
+			for i, sco := range txn.SiacoinOutputs {
+				// if a seed output is found, add it to the output set
+				if index, exists := s.keys[sco.UnlockHash]; exists && sco.Value.Cmp(s.dustThreshold) > 0 {
+					id := types.OutputID(txn.SiacoinOutputID(uint64(i)))
+					s.siacoinOutputs[id] = scannedOutput{
+						id:        id,
+						value:     sco.Value,
+						seedIndex: index,
+					}
+				}
+			}
+			// TODO: should this be done in a separate txn loop? This would be
+			// mandatory if txn0 can spend an output created in txn1
+			for _, sci := range txn.SiacoinInputs {
+				// if a seed output is spent, remove it from the output set
+				id := types.OutputID(sci.ParentID)
+				if _, exists := s.siacoinOutputs[id]; exists {
+					delete(s.siacoinOutputs, id)
+				}
+			}
+		}
+	}
+
+	// update s.largestIndexSeen
+	var addrs []types.UnlockHash
+	for _, diff := range cc.SiacoinOutputDiffs {
+		addrs = append(addrs, diff.SiacoinOutput.UnlockHash)
+	}
+	for _, diff := range cc.SiafundOutputDiffs {
+		addrs = append(addrs, diff.SiafundOutput.UnlockHash)
+	}
+	for _, block := range cc.AppliedBlocks {
+		for _, mp := range block.MinerPayouts {
+			addrs = append(addrs, mp.UnlockHash)
+		}
+		for _, txn := range block.Transactions {
+			for _, sci := range txn.SiacoinInputs {
+				addrs = append(addrs, sci.UnlockConditions.UnlockHash())
+			}
+			for _, sco := range txn.SiacoinOutputs {
+				addrs = append(addrs, sco.UnlockHash)
+			}
+			for _, sfi := range txn.SiafundInputs {
+				addrs = append(addrs, sfi.UnlockConditions.UnlockHash())
+			}
+			for _, sfo := range txn.SiafundOutputs {
+				addrs = append(addrs, sfo.UnlockHash)
+			}
+		}
```

Comment:
all of these unlock hashes should be appearing in cc.SiacoinOutputDiffs, I don't think you need to do this extra scanning.